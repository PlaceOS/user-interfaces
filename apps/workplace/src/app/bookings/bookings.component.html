<ng-container [ngSwitch]="flow">
    <app-desk-flow (menu)="show_menu = $event" *ngSwitchCase="'desks-map'"></app-desk-flow>
    <desk-flow-listing (menu)="show_menu = $event" *ngSwitchCase="'desks'"></desk-flow-listing>
    <space-flow
        (menu)="show_menu = $event"
        *ngSwitchDefault
    ></space-flow>
    <qr-reader
        *ngSwitchCase="'qr'"
        (menu)="show_menu = $event"
    >
    </qr-reader>

</ng-container>
<a-overlay-menu [(show)]="show_menu"></a-overlay-menu>
