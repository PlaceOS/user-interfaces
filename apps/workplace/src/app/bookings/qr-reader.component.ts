import { Component, OnInit, Output, EventEmitter, ViewChild, ElementRef, OnDestroy } from '@angular/core';
import { Router } from '@angular/router';
import { BrowserQRCodeReader } from '@zxing/library';
import { BaseClass } from '@placeos/common';

@Component({
    selector: 'qr-reader',
    template: `
        <header class="header">
            <a-topbar-header (menuChange)="menu.emit($event)"></a-topbar-header>
        </header>
        <div
            class="flex-1 overflow-hidden flex items-center justify-center bg-black relative"
        >
            <video
                class="min-w-full min-h-full object-cover"
                id="video"
            ></video>
            <div
                class="absolute text-white text-center inset-0 flex flex-col items-center justify-center"
            >
                <div
                    class="flex flex-col z-10 justify-end items-center relative"
                    *ngIf="is_scanning"
                >
                    <h2 class="text-3xl mb-1 uppercase subpixel-antialiased">
                        Scan QR Code
                    </h2>
                    <span class=" mb-4 ">
                        Scan the QR code outisde a PlaceOS room or space.
                    </span>
                </div>
                <div
                    class="flex flex-col z-10 justify-end items-center relative"
                    *ngIf="!is_scanning"
                >
                    <h2 class="text-3xl mb-1 uppercase subpixel-antialiased">
                        Enter Room ID
                    </h2>
                    <span class="mb-4">
                        Enter the room ID number outisde a PlaceOS room or
                        space.
                    </span>
                </div>
                <div class="flex items-center justify-center">
                    <div
                        box
                        class="rounded-2xl h-64 w-64 transition-all flex items-center justify-center p-8 m-8 space-x-2"
                        [class.input]="!is_scanning"
                    >
                        <span class="uppercase">Booking ID</span>
                        <input
                            matInput
                            class="border-none bg-none w-full text-3xl text-left"
                            [(ngModel)]="room_code"
                            name="booking-id"
                            placeholder="e.g. 12102910"
                        />
                    </div>
                </div>
                <div
                    class="m-4 p-2 flex items-center space-x-2 bg-white bg-opacity-50 rounded"
                >
                    <button
                        mat-button
                        [class]="
                            'flex-1 text-black border-none w-40 ' +
                            (is_scanning
                                ? 'bg-white'
                                : 'hover:bg-white bg-opacity-50')
                        "
                        (click)="is_scanning = true"
                    >
                        Scan Code
                    </button>
                    <button
                        mat-button
                        [class]="
                            'flex-1 text-black border-none w-40 ' +
                            (!is_scanning
                                ? 'bg-white'
                                : 'hover:bg-white bg-opacity-50')
                        "
                        (click)="is_scanning = false"
                    >
                        Enter Code
                    </button>
                </div>
            </div>
        </div>
    `,
    styles: [
        `
            :host {
                width: 100%;
                height: 100vh;
                display: flex;
                flex-direction: column;
            }

            h2 {
                font-family: 'Grotesk Condensed';
            }

            [box] {
                box-shadow: 0px 0px 0px 100vw rgba(0, 0, 0, 0.5);
            }

            [box] > * {
                display: none;
            }

            [box].input {
                width: 32rem !important;
                max-width: calc(100% - 2rem) !important;
                padding: 1rem !important;
                height: 4rem !important;
                color: black !important;
                background: white;
                box-shadow: 0px 0px 0px 100vw rgba(0, 0, 0, 0.8);
            }

            [box].input > * {
                display: initial;
            }

            [box] span {
                font-family: 'Grotesk Bold';
                text-transform: uppercase;
                letter-spacing: 0.05em;
            }

            button {
                background: transparent;
            }
        `,
    ],
})
export class QrReaderComponent extends BaseClass implements OnInit, OnDestroy {
    /** Whether to show the overlay menu */
    public show_menu: boolean;
    /** Boolean to toggle scan/code */
    public is_scanning: boolean = true;
    /** Room Code input value */
    public room_code: string;
    /** Menu event */
    @Output() public menu = new EventEmitter(false);
    /** Device ID of Camera */
    private device_id: string;
    /** Video element to emit camera feed */
    @ViewChild('video') private _video_el: ElementRef<HTMLVideoElement>;

    constructor(private _router: Router) {
        super();
    }

    public ngOnDestroy() {
        if (this._video_el?.nativeElement.srcObject) {
            const track = (this._video_el.nativeElement
                .srcObject as any).getTracks()[0];
            if (track) {
                track.stop();
            }
        }
    }

    public ngOnInit(): void {
        const codeReader = new BrowserQRCodeReader();
        codeReader.listVideoInputDevices().then((videoInputDevices) => {
            this.device_id = videoInputDevices[0].deviceId;
            codeReader
                .decodeOnceFromVideoDevice(this.device_id, 'video')
                .then((result): any => {
                    const url = result['text'];
                    const hashindex = url.indexOf('/#/');
                    // could just whole page redirect to url
                    if (hashindex > 0) {
                        let regex = /[?&]([^=#]+)=([^&#]*)/g,
                            params = {},
                            match;
                        while ((match = regex.exec(url))) {
                            params[match[1]] = match[2];
                        }
                        this._router.navigate(
                            [url.split('/#')[1].split('?')[0]],
                            params
                        );
                    }
                })
                .catch((err) => {
                    console.error(err);
                });
        });
    }
}
