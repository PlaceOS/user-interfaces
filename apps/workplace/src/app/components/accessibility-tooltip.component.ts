import { Component, Inject } from '@angular/core';
import { ApplicationLink, SettingsService } from '@placeos/common';
import { CustomTooltipData } from '@placeos/components';

@Component({
    selector: 'accessibility-tooltip',
    template: `
        <div
            class="flex flex-col w-[18.5rem] rounded bg-white shadow relative -top-12 -right-1 overflow-hidden dark:bg-neutral-700 dark:text-white"
        >
            <div class="flex items-center space-x-2 p-2" (click)="close()">
                <app-icon class="text-2xl">arrow_back</app-icon>
                <div class="" i18n>Display & Accessibility</div>
            </div>
            <div
                action
                class="flex items-center space-x-2 w-full text-left h-auto px-4"
            >
                <div
                    class="flex items-center justify-center rounded-full w-8 h-8 bg-gray-200 dark:bg-neutral-800 my-2"
                >
                    <app-icon class="text-xl">mode_night</app-icon>
                </div>
                <div class="flex-1 text-sm" i18n>Dark Mode</div>
                <mat-slide-toggle
                    [ngModel]="dark_mode"
                    (ngModelChange)="applySetting('dark_mode', $event)"
                ></mat-slide-toggle>
            </div>
            <div
                action
                class="flex items-center space-x-2 w-full text-left h-auto px-4"
            >
                <div
                    class="flex items-center justify-center rounded-full w-8 h-8 bg-gray-200 dark:bg-neutral-800 my-2"
                >
                    <app-icon class="text-xl">playlist_add</app-icon>
                </div>
                <div class="flex-1 text-sm" i18n>Large Accessibility Sizes</div>
                <mat-slide-toggle
                    [ngModel]="accessible"
                    (ngModelChange)="applySetting('accessible', $event)"
                ></mat-slide-toggle>
            </div>
            <ng-container *ngIf="accessible">
                <div class="px-8 py-4 bg-gray-200 dark:bg-neutral-600 text-center" i18n>
                    Adjust your preferred reading size below.
                </div>
                <div class="flex items-center px-4 space-x-4">
                    <span class="text-sm">A</span>
                    <mat-slider
                        class="flex-1 w-1/2"
                        [min]="10"
                        [max]="22"
                        [ngModel]="font_size"
                        (ngModelChange)="applySetting('font_size', $event)"
                        [step]="2"
                    ></mat-slider>
                    <span class="text-2xl">A</span>
                </div>
            </ng-container>
        </div>
    `,
    styles: [
        `
            :host > div > [action] {
                border-top: 1px solid #ccc;
            }
        `,
    ],
})
export class AccessibilityTooltipComponent {

    public get dark_mode() {
        return !!this._settings.get('dark_mode');
    }

    public get accessible() {
        return !!this._settings.get('accessible');
    }

    public get font_size() {
        return this._settings.get('font_size') || 16;
    }

    public readonly applySetting = (n, v) =>
        this._settings.saveUserSetting(n, v);

    public readonly close = () => this._data?.close();

    constructor(
        @Inject(CustomTooltipData) private _data: any,
        private _settings: SettingsService
    ) {}
}
