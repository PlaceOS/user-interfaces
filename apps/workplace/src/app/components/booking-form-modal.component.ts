import { Component, Input, OnInit } from '@angular/core';
import { FormGroup } from '@angular/forms';

import { BaseClass } from '@placeos/common';

@Component({
    selector: 'a-booking-form-modal',
    template: `
        <div class="booking-form" [formGroup]="form">
            <div *ngIf="first_form">
                <div>
                    <label>
                        <strong>Meeting Subject:<span>*</span></strong>
                    </label>
                    <mat-form-field appearance="outline">
                        <input
                            matInput
                            name="title"
                            placeholder="Enter here"
                            formControlName="title"
                        />
                    </mat-form-field>
                </div>
                <div *ngIf="form.controls.organiser">
                    <label for="organiser">
                        <strong>Host:<span>*</span></strong>
                    </label>
                    <a-user-search-field
                        name="organiser"
                        formControlName="organiser"
                    ></a-user-search-field>
                </div>
                <div>
                    <label for="start-time">
                        <strong>Time:<span>*</span></strong>
                    </label>
                    <a-time-field
                        name="start-time"
                        formControlName="date"
                    ></a-time-field>
                </div>
                <div style="margin-top: 1.5em;">
                    <label for="duration">
                        <strong>Duration:<span>*</span></strong>
                    </label>
                    <a-duration-field
                        name="duration"
                        formControlName="duration"
                    ></a-duration-field>
                </div>
            </div>
            <div [hidden]="first_form">
                <input
                    *ngIf="form.controls.notes"
                    class="notes"
                    matInput
                    name="notes"
                    placeholder="Add notes here..."
                    formControlName="notes"
                />
                <div *ngIf="form.controls.head_count">
                    <label for="organiser">
                        <strong>Estimated Headcount:<span>*</span></strong>
                    </label>
                    <mat-form-field appearance="outline">
                        <input
                            matInput
                            name="headcount"
                            placeholder="Expected attendees here.."
                            formControlName="head_count"
                        />
                    </mat-form-field>
                </div>
            </div>
        </div>
    `,
    styles: [
        `
            .field {
                display: flex;
                flex-wrap: wrap;
            }

            .notes {
                padding-bottom: 150px;
                width: 100%;
                background-color: lightgrey;
                border: solid 1px rgba(black, 0.5);
                margin-bottom: 50px;
            }

            mat-form-field {
                width: 100%;
            }
        `,
    ],
})
export class BookingFormModalComponent extends BaseClass {
    @Input() public first_form: boolean;
    @Input() public form: FormGroup;
}
