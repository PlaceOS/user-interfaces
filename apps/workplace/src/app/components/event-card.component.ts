import { Component, Input } from '@angular/core';
import { CalendarEvent } from '@placeos/events';
import { addMinutes, format, formatDuration, isSameDay } from 'date-fns';

@Component({
    selector: 'event-card',
    template: `
        <h4 class="mb-2 flex items-center" *ngIf="event" date>
            <span *ngIf="show_day" day>{{ day }}, </span>
            {{ event?.date | date: 'h:mm a' }}
            <span class="text-xs px-2">({{ event?.date | date: 'z' }})</span>
        </h4>
        <a
            matRippleColor
            details
            class="w-full"
            [routerLink]="['/schedule', 'view', event?.id, 'event']"
            *ngIf="event"
        >
            <div class="w-full bg-white rounded shadow py-2 relative">
                <h4 class="px-2">{{ event?.title }}</h4>
                <div class="flex m-2">
                    <div
                        class="flex items-center bg-opacity-30 rounded-2xl p-1 text-sm space-x-2 pr-2 font-medium"
                        [class.bg-green-600]="event?.status === 'approved'"
                        [class.bg-yellow-500]="event?.status === 'tentative'"
                        [class.bg-red-600]="event?.status === 'declined'"
                        [class.bg-gray-200]="!event?.status"
                    >
                        <div
                            class="rounded-full h-5 w-5 flex items-center justify-center text-white"
                            [class.bg-success]="event?.status === 'approved'"
                            [class.bg-pending]="event?.status === 'tentative'"
                            [class.bg-error]="event?.status === 'declined'"
                            [class.bg-gray-200]="!event?.status"
                        >
                            <app-icon>
                                {{
                                    event?.status === 'approved'
                                        ? 'done'
                                        : event?.status === 'tentative'
                                        ? 'warning'
                                        : 'close'
                                }}
                            </app-icon>
                        </div>
                        <div class="pr-1">{{ period }}</div>
                    </div>
                </div>
                <div
                    class="flex flex-wrap flex-col sm:flex-row sm:divide-x divide-gray-200 py-2 space-y-2 sm:space-y-0"
                >
                    <div class="flex items-center px-2">
                        <app-icon>meeting_room</app-icon>
                        <div class="mx-2 truncate">
                            {{
                                event?.space?.level?.display_name ||
                                    event?.space?.level?.name
                            }},
                            {{
                                event?.space?.display_name || event?.space?.name
                            }}
                        </div>
                    </div>
                    <div class="flex items-center px-2" *ngIf="location">
                        <app-icon>place</app-icon>
                        <div class="mx-2 truncate">{{ location }}</div>
                    </div>
                    <div class="flex items-center px-2">
                        <app-icon>person_outline</app-icon>
                        <div class="mx-2">
                            {{
                                event?.organiser?.name ||
                                    event?.organiser?.email
                            }}
                        </div>
                    </div>
                    <div
                        class="flex items-center px-2"
                        *ngIf="event?.ext('catering')?.length"
                    >
                        <app-icon>restaurant</app-icon>
                        <div class="mx-2">Catered</div>
                    </div>
                    <div class="flex items-center px-2">
                        <app-icon>people</app-icon>
                        <div class="mx-2">
                            {{ event?.attendees?.length }}
                            {{
                                event?.attendees?.length === 1
                                    ? 'Person'
                                    : 'People'
                            }}
                        </div>
                    </div>
                </div>
                <app-icon
                    class="absolute top-1/2 right-1 text-4xl -translate-y-1/2"
                >
                    chevron_right
                </app-icon>
                <div
                    class="absolute bottom-2 right-2 sm:bottom-auto sm:top-2 text-sm sm:text-base flex items-center pr-4"
                    *ngIf="event?.attendees?.length"
                >
                    <div
                        class="h-10 w-6"
                        *ngFor="
                            let user of event?.attendees
                                | slice
                                    : 0
                                    : (event?.attendees?.length === 6 ? 6 : 5)
                        "
                    >
                        <a-user-avatar [user]="user"></a-user-avatar>
                    </div>
                    <div class="h-10 w-6" *ngIf="event?.attendees?.length > 6">
                        <div
                            class="bg-secondary rounded-full h-10 w-10 border-2 border-white flex items-center justify-center text-white"
                        >
                            +{{ event?.attendees?.length - 5 }}
                        </div>
                    </div>
                </div>
            </div>
        </a>
    `,
    styles: [
        `
            :host {
                display: block;
                width: 100%;
            }
        `,
    ],
})
export class EventCardComponent {
    @Input() public event: CalendarEvent;
    @Input() public show_day: boolean = false;

    public get day() {
        const date = this.event?.date || Date.now();
        const is_today = isSameDay(Date.now(), date);
        return `${is_today ? 'Today' : format(date, 'EEEE')}`;
    }

    public get location() {
        return '';
    }

    public get period() {
        const start = this.event?.date || Date.now();
        const duration = this.event?.duration || 60;
        const end = addMinutes(start, duration);
        const dur = formatDuration({
            hours: Math.floor(duration / 60),
            minutes: duration % 60,
        })
            .replace(' hour', 'hr')
            .replace(' minute', 'min');
        return `${format(start, 'h:mm a')} - ${format(end, 'h:mm a')} (${dur})`;
    }
}
