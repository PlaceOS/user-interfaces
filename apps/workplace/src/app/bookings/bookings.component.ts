import { Component, OnInit } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';

import { clearBookingFormData, clearDeskBookingFormData } from '@user-interfaces/bookings';
import { BaseClass, HashMap, SettingsService } from '@user-interfaces/common';


export interface BookingFlowEvent {
    /** Type of flow event */
    type: 'previous' | 'next' | 'reset' | 'close' | 'error';
    /** Step which the event was emitted from */
    step: string;
    /** Additional metadata for the event */
    metadata?: HashMap;
}

@Component({
    selector: 'app-bookings',
    templateUrl: './bookings.component.html',
    styleUrls: ['./bookings.component.scss'],
})
export class BookingsComponent extends BaseClass implements OnInit {
    /** Booking flow */
    public flow: 'spaces' | 'desks' | 'qr' | '';
    /** Whether to show the overlay menu */
    public show_menu: boolean;

    constructor(
        private _route: ActivatedRoute,
        private _router: Router,
        private _settings: SettingsService
    ) {
        super();
    }

    public ngOnInit(): void {
        this._settings.title = 'Book';
        this.subscription(
            'route.params',
            this._route.paramMap.subscribe((params) => {
                if (params.has('flow')) {
                    this.flow = params.get('flow') as any;
                } else {
                    this.flow = '';
                }
            })
        );
        this.subscription(
            'query.params',
            this._route.queryParamMap.subscribe((params) => {
                if (params.has('new')) {
                    clearDeskBookingFormData();
                    clearBookingFormData();
                    this._router.navigate([], {
                        queryParams: {
                            new: null,
                        },
                        replaceUrl: true,
                        queryParamsHandling: 'merge',
                    });
                }
            })
        );
    }

    public spaces() {
        this._router.navigate(['spaces'], { relativeTo: this._route });
    }
}
