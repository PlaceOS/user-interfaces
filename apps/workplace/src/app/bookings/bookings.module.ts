import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule, Route } from '@angular/router';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { BookingsComponent } from './bookings.component';
import { SharedContentModule } from '../ui/shared.module';
import { SharedExploreModule } from '@user-interfaces/explore';
import { SharedUsersModule } from '@user-interfaces/users';

import { QrReaderComponent } from './qr-reader/qr-reader.component';

import { SpaceFlowComponent } from './space-flow/space-flow.component';
import { SpaceFlowFormComponent } from './space-flow/space-flow-form.component';
import { SpaceFlowListingComponent } from './space-flow/space-flow-listing.component';
import { SpaceFlowConfirmComponent } from './space-flow/space-flow-confirm.component';

import { DeskFlowComponent } from './desk-flow/desk-flow.component';
import { DeskFlowListingComponent } from './desk-flow/desk-flow-listing.component';
import { DeskFlowConfirmModalComponent } from './desk-flow/desk-flow-confirm-modal.component';
import { DeskFlowQuestionsModalComponent } from './desk-flow/desk-flow-questions-modal.component';

const ROUTES: Route[] = [
    { path: ':flow', component: BookingsComponent },
    { path: ':flow/:step', component: BookingsComponent },
    { path: '', redirectTo: 'spaces' },
];

@NgModule({
    declarations: [
        BookingsComponent,

        DeskFlowComponent,
        DeskFlowListingComponent,
        DeskFlowConfirmModalComponent,
        DeskFlowQuestionsModalComponent,

        SpaceFlowComponent,
        SpaceFlowFormComponent,
        SpaceFlowListingComponent,
        SpaceFlowConfirmComponent,

        QrReaderComponent,
    ],
    imports: [
        CommonModule,
        RouterModule.forChild(ROUTES),
        FormsModule,
        ReactiveFormsModule,
        SharedContentModule,
        SharedUsersModule,
        SharedExploreModule
    ],
})
export class BookingsModule {

}
