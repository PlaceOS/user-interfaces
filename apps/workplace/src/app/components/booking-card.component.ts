import { Component, Input } from '@angular/core';
import { Booking } from '@placeos/bookings';
import { OrganisationService } from '@placeos/organisation';
import { addMinutes, format, formatDuration, isSameDay } from 'date-fns';

@Component({
    selector: 'booking-card',
    template: `
        <h4 class="mb-2 flex items-center" *ngIf="booking">
            <span *ngIf="show_day">{{ day }}, </span>
            {{ booking?.date | date: 'h:mm a' }}
            <span class="text-xs px-2">({{ booking?.date | date:'z' }})</span>
        </h4>
        <a
            matRippleColor
            class="w-full"
            [routerLink]="['/schedule', 'view', booking?.id, 'booking']"
            *ngIf="booking"
        >
            <div class="w-full bg-white rounded shadow py-2 relative">
                <h4 class="px-2">{{ booking?.title }}</h4>
                <div class="flex m-2">
                    <div
                        class="flex items-center bg-opacity-30 rounded-2xl p-1 text-sm space-x-2 pr-2 font-medium"
                        [class.bg-success]="booking?.status === 'approved'"
                        [class.bg-pending]="booking?.status === 'tentative'"
                        [class.bg-error]="booking?.status === 'declined'"
                        [class.bg-gray-200]="!booking?.status"
                    >
                        <div
                            class="rounded-full h-5 w-5 flex items-center justify-center text-white"
                            [class.bg-success]="booking?.status === 'approved'"
                            [class.bg-pending]="booking?.status === 'tentative'"
                            [class.bg-error]="booking?.status === 'declined'"
                            [class.bg-gray-200]="!booking?.status"
                        >
                            <app-icon>
                                {{
                                    booking?.status === 'approved'
                                        ? 'done'
                                        : booking?.status === 'tentative'
                                        ? 'warning'
                                        : 'close'
                                }}
                            </app-icon>
                        </div>
                        <div class="pr-1">{{ period }}</div>
                    </div>
                </div>
                <div
                    class="flex flex-wrap flex-col sm:flex-row sm:divide-x divide-gray-200 py-2 space-y-2 sm:space-y-0"
                >
                    <div class="flex items-center px-2">
                        <app-icon *ngIf="type !== 'desk'; else desk_icon">{{ type }}</app-icon>
                        <div class="mx-2 truncate">
                            {{ booking?.description || booking?.asset_id }}
                        </div>
                    </div>
                    <div class="flex items-center px-2" *ngIf="location">
                        <app-icon>place</app-icon>
                        <div class="mx-2 truncate">{{ location }}</div>
                    </div>
                </div>
                <app-icon
                    class="absolute top-1/2 right-1 text-4xl -translate-y-1/2"
                >
                    chevron_right
                </app-icon>
            </div>
        </a>
        <ng-template #desk_icon>
            <img src="assets/icons/desk-outline.svg" class="w-5 h-5" />
        </ng-template>
    `,
    styles: [
        `
            :host {
                display: block;
                width: 100%;
            }
        `,
    ],
})
export class BookingCardComponent {
    @Input() public booking: Booking;
    @Input() public show_day: boolean = false;

    public get type() {
        if (this.booking?.type === 'desk') return 'desk';
        if (this.booking?.type === 'parking') return 'drive_eta';
        return 'book';
    }

    public get day() {
        const date = this.booking?.date || Date.now();
        const is_today = isSameDay(Date.now(), date);
        return `${is_today ? 'Today' : format(date, 'EEEE')}`;
    }

    public get location() {
        const level = this._org.levelWithID(this.booking?.zones || []);
        return `${level?.display_name || level?.name || ''}`;
    }

    public get period() {
        const start = this.booking?.date || Date.now();
        const duration = this.booking?.duration || 60;
        const end = addMinutes(start, duration);
        const dur = formatDuration({
            hours: Math.floor(duration / 60),
            minutes: duration % 60,
        })
            .replace(' hour', 'hr')
            .replace(' minute', 'min');
        return `${format(start, 'h:mm a')} - ${format(end, 'h:mm a')} (${dur})`;
    }

    constructor(private _org: OrganisationService) {}
}
