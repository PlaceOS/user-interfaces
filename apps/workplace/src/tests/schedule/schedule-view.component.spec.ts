import { FormsModule } from '@angular/forms';
import { MatDialog } from '@angular/material/dialog';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { Router } from '@angular/router';
import { createRoutingFactory, SpectatorRouting } from '@ngneat/spectator/jest';
import { IconComponent, UserAvatarComponent } from '@placeos/components';
import { EventFormService } from '@placeos/events';
import { MockComponent } from 'ng-mocks';

import { CalendarEvent } from 'libs/events/src/lib/event.class';

import { ScheduleViewComponent } from '../../app/schedule/schedule-view.component';

jest.mock('@placeos/common');
jest.mock('@placeos/bookings');
jest.mock('@placeos/events');

import * as common_mod from '@placeos/common';
import * as booking_mod from '@placeos/bookings';
import * as events_mod from '@placeos/events';
import { of } from 'rxjs';

describe('ScheduleViewComponent', () => {
    let spectator: SpectatorRouting<ScheduleViewComponent>;
    const createComponent = createRoutingFactory({
        component: ScheduleViewComponent,
        providers: [
            { provide: MatDialog, useValue: { open: jest.fn() } },
            { provide: EventFormService, useValue: { newForm: jest.fn() } },
        ],
        imports: [FormsModule, MatProgressSpinnerModule],
        declarations: [
            MockComponent(IconComponent),
            MockComponent(UserAvatarComponent),
        ],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });

    it('should allow viewing locations of spaces', () => {
        spectator.detectChanges();
        spectator.component.event = new CalendarEvent({
            system: { id: '1', email: '1', resource: true } as any,
        });
        spectator.detectChanges();
        expect('button[locate]').toExist();
        spectator.click('button[locate]');
        expect(spectator.inject(MatDialog).open).toHaveBeenCalled();
    });

    it('should allow editing of events', () => {
        spectator.detectChanges();
        spectator.component.event = new CalendarEvent({
            system: { id: '1', email: '1' } as any,
        });
        spectator.detectChanges();
        expect('button[edit]').toExist();
        spectator.click('button[edit]');
        expect(spectator.inject(EventFormService).newForm).toHaveBeenCalled();
        expect(spectator.inject(Router).navigate).toHaveBeenCalledWith([
            '/book',
            'spaces',
            'form',
        ]);
    });

    it('should allow deleting events and bookings', () => {
        spectator.component.event = new CalendarEvent({
            resources: [{ id: '1' } as any],
        });
        spectator.detectChanges();
        expect('button[remove]').toExist();
        (common_mod as any).openConfirmModal = jest.fn(() => ({
            reason: 'done',
            loading: jest.fn(),
        }));
        (events_mod as any).removeEvent = jest.fn(() => of({}));
        spectator.click('button[remove]');
        expect(common_mod.openConfirmModal).toHaveBeenCalledTimes(1);
        expect(spectator.inject(Router).navigate).toHaveBeenCalledWith([
            '/schedule',
        ]);
    });
});
