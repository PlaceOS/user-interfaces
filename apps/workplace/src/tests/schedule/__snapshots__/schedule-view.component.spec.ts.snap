// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScheduleViewComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    topbar=""
  >
    <a
      button=""
      class="clear h-[2.75rem] text-white"
      mat-button=""
    >
      <div
        class="flex items-center justify-center h-full"
      >
        <app-icon
          class="text-xl"
        >
          arrow_back
        </app-icon>
        <span
          class="ml-2 mx-4"
        >
          Back
        </span>
      </div>
    </a>
  </div>
  <div
    class="flex-1 w-full flex flex-col items-center bg-gray-200 p-4 overflow-auto"
  >
    <div
      class="h-full w-full flex flex-col items-center justify-center space-y-2"
    >
      <mat-spinner
        class="mat-spinner mat-progress-spinner mat-primary mat-progress-spinner-indeterminate-animation _mat-animation-noopable"
        mode="indeterminate"
        role="progressbar"
        style="width: 32px; height: 32px;"
      >
        <svg
          aria-hidden="true"
          focusable="false"
          preserveAspectRatio="xMidYMid meet"
          style="width: 32px; height: 32px;"
          viewBox="0 0 25.2 25.2"
        >
          <circle
            cx="50%"
            cy="50%"
            r="11"
            style="animation-name: mat-progress-spinner-stroke-rotate-32; stroke-dasharray: 69.11503837897544px; stroke-width: 10%;"
          />
          
          
        </svg>
      </mat-spinner>
      <p>
        Loading event data...
      </p>
    </div>
    
  </div>
  
</div>
`;
